### ISSUE TITLE
[Feature] Hibernate

### ISSUE THREAD

**Issue by m1k1o** [m1k1o/neko#57]
Considering, if it would be interensting, to have some kind of hibernate feature. It has already been discussed, that browser does not need to be running always, when neko is idle. It could "sleep" or just be hibernating. Meaning, Go serwer would run but all the pipelines (maybe xorg & pulseaudio) and browser would be shut down.

There would be a button (only for admins) that can "wake it up". Others could connect to neko, but they would be notified about it is "sleeping".

This would introduce much refactoring on the connectionl logic on client / server. Considering, if it is worth it.

**Comment by mbattista** [comment]
If I understand correctly you just want a possibility to stop/start the gstreamer pipeline on command?

**Comment by m1k1o** [comment]
Not only gstreamer pipeline but also browser and WebRTC server.

**Comment by mbattista** [comment]
Why also browser and WebRTC?

Stopping the pipeline would leave neko in the `starting` state, before anyone has joined and could save some cpu cores.
But the browser and the WebRTC server do not need that many resources?

If you e.g. stop an youtube video to watch later and hypernate kills the browser, you have to remember how far along the video you were.

**Comment by m1k1o** [comment]
My motivation is to be able to disconnect users mainly from streaming video from my server, while being connected to chat. So you are right, no need to turn off browser.

If I have my deployment and I have some people, they are constantly pulling video from my server. Unless I turn off neko, they disconnect or I kick them. I thought about having some lobby, so people can be connected and chat while not receiving video data from server unless admin allows.

Another level of this would be to have this split between users and admins. So allow only admins to stream (logging in to accounts, typing some sensitive information) while users would wait in the lobby and after they will be let in.

**Comment by m1k1o** [comment]
I like the first solution, and I think, it could be indeed sufficient. That would create a new usecase: Truning only video pipeline off, so if you are just listening to music, you don't need to receive video stream.

**Comment by mbattista** [comment]
This feature took longer then I expected and is still not finished.

It will still need a lot of re-code since it will now need to save all the peerconnections and corresponding websockets to reanounce SDPs after each change of track (removed audio / video). I will try to tackle #48 while I am at it.

**Comment by m1k1o** [comment]
You can open a draft pull request and I can help with some tasks, if you want.

**Comment by Kinuseka** [comment]
I wonder if this is still being worked on? Its a very helpful feature.

**Comment by m1k1o** [comment]
I don't think this is being worked on. It can be acheived by stopping `docker exec neko supervisorctl stop google-chrome` and starting `docker exec neko supervisorctl start google-chrome`. You could wrap it to a simple service and create custom button for that.

**Comment by Kinuseka** [comment]
I was hoping for an automatic hibernation there are no more users on the room.

**Comment by m1k1o** [comment]
That info can be received using `/stats?pwd=<admin>` endpoint. But yeah, it would be nicer if it would be part of neko itself. But the project is getting bigger and bigger, maybe it would be a good time to think about some plugins/addons. Where such logic could be easily added to neko externally.
